### ISSUE TITLE
TA and Coloring is very wrong

### ISSUE THREAD

**Issue by dozer8383** [Despatra/Phoxel#14]
**Describe the bug**
Somewhat creepy lines appear all over the screen and grow in thickness as I look around, until the whole world goes totally black. The whole world, when not being obscured by these lines, is cast in a dark cyan/green color, where only the edges of the blocks have a lot of noise.

**Steps to Reproduce**
n/a

**Screenshots**
![2024-07-02_18 55 17](https://github.com/Despatra/Phoxel/assets/119268610/033d7f30-c561-4c70-9e25-365eaa5cd342)
![2024-07-02_18 56 26](https://github.com/Despatra/Phoxel/assets/119268610/b1ff7c5b-f985-4d25-84c5-f16978b9338e)
![2024-07-02_18 56 05](https://github.com/Despatra/Phoxel/assets/119268610/269efb59-0cc3-4fb7-8ba9-8755e6004902)
![2024-07-02_18 55 40](https://github.com/Despatra/Phoxel/assets/119268610/153afeb8-8a46-4409-8368-46299b6f1730)
**Specifications if it has to do with performance or crashes**
In this case it's not really performance nor crashes, but maybe you'll find it helpful. I'm getting about 35 FPS as it is right now, which is great to me.
 - Graphics Card: Integrated Intel graphics (CPU is 11th Gen Intel® Core™ i5-1145G7)
 - Minecraft Version: 1.20.4
 - Shader Version: 0.4.0.3 (Beta 0.4 Patch 3)
 - Shader Loader: Iris (with Sodium, if that makes a difference)

**Additional context**
On Ubuntu 24.04. Previous versions of Phoxel (last version tested was 0.4 Patch 2) would simply cause Iris to say "The shaderpack failed to load!"...

Also unrelated to this bug report, I really love the idea of this shader and can't wait to see it working on my machine!

**Comment by Despatra** [comment]
This is extremely interesting... 
The lines on borders when you move is just an artifact of temporal accumulation, as it has to reset the data in new areas being shown, not much can be done about that.
But the fact that it's so bright and so green is definitely confusing. the whole image seems to be a lot more green than it should be.
The only thing I can think of is that it's integrated, I really haven't seen anything like this...

**Comment by dozer8383** [comment]
Update: 
I tried setting Temporal Accumulation to OFF, and now everything is bright white except for the sky which appears moderately normal, and a small band at the bottom of the screen which just appears like an insanely oversaturated photo. When I press F1 to turn off the GUI, the white begins to form weird patterns with yellow, blue, and green, depending on the camera angle, and turning the GUI back on seems to pause it from changing colors. I tried turning Temporal Accumulation back to its default value and to a high value and now it won't go back to the green tinted, growing black lines phenomenon.
![2024-07-05_08 56 21](https://github.com/Despatra/Phoxel/assets/119268610/f5da2387-83ce-4bb5-8179-6c9a56761f01)
![2024-07-05_08 56 34](https://github.com/Despatra/Phoxel/assets/119268610/c5d60a81-905e-48b1-a829-26479bf42dc9)

**Comment by ShrikerChestTrees** [comment]
> Hah. Yeah, integrated graphics hardly play nicely with anything, but I've never had any problems with other shaders aside from just low FPS on the high settings. Oh well. Something about 0.4 Patch 3 made the shader load this time, all the other versions won't load, so you changed something that made it work (though, clearly, not correctly). Keep up the good work, maybe it'll fix itself.

Strangely everything works fine for me on an Intel UHD Graphics integrated graphics card...

**Comment by Despatra** [comment]
Hmm, yeah this shader seems kind of hit and miss with some people.

**Comment by dozer8383** [comment]
Update for 0.4 Patch 4:
So, things are better but also not. No big bleeding black lines any more, but colors are still messed up. These are screenshots from a world (modded). One is without shaders and the other Phoxel. If I turn off the GUI when Phoxel is enabled, everything goes black, and while you can see the GUI fine if you enable it again, the world stays black. Only turning shaders off and back on will make it work again.
Not sure if this will help at all, but thought I'd note it anyway.

![2024-08-29_19 15 20](https://github.com/user-attachments/assets/f2dd9fd7-39a1-4649-923d-7451846b1d85)
![2024-08-29_19 15 13](https://github.com/user-attachments/assets/a6be4435-97b7-4d7e-b126-e0c29bc7eac6)

**Comment by Despatra** [comment]
It's almost like it's corrupted... I checked your IGPU's RAM and perhaps you don't have enough?, I feel as though it would just crash, but I don't know. I found that your IGPU is limited to 1/2 your systems memory, but it doesn't seem to be set to that on default. The only way to tell is to check your VRAM usage.

**Comment by dozer8383** [comment]
So, I realized that 1.20.1 is not supported by patch 4 or 0.5.1, so now here's more updates, with the F3 screen open so you can see the system stats. It actually looks OK now. These screenshots are on 1.21.1 with Phoxel 0.5.1.
![2024-09-02_16 21 15](https://github.com/user-attachments/assets/d15aefe2-66d3-482a-bc48-a5d6eea44737)
![2024-09-02_16 22 12](https://github.com/user-attachments/assets/07f1a333-b425-4863-9e12-da042e1a52ec)
It looked fine, no severe discoloration (but no pathtracing goodness still), until I turned on panorama mode.
![2024-09-02_16 31 24](https://github.com/user-attachments/assets/eb789e83-a9c8-4594-bacd-0eff7889df3b)
![2024-09-02_16 31 41](https://github.com/user-attachments/assets/350b58af-6f90-44a0-99ff-eea515f30ad0)
Putting it back to normal did not revert this.
![2024-09-02_16 32 59](https://github.com/user-attachments/assets/180c0b30-7d75-4972-b11a-7518ff787889)
It only got worse with time until it was basically a seizure waiting to happen.
Thanks for all your support and work in investigating this.

**Comment by Despatra** [comment]
Ok, so few helpful things I've noticed, which I'm going to mark here.
1. Voxelization does work (blocks and their textures work in image 4)
2. The PT tracing code itself seems to work as well (shadows and GI in image 4 & 5)
3. From the artifacts in the first 2 images parallax (pixel lines on block edges) and "transparent" water (noisy water) is working
This means that the issue is likely from:
* composite2.fsh
* corrupted texture formats (likely which ones store the PT data being RGBA16-32F, most likely the RGBA32F)\

I will send a modified pack when I can with these changed in hopes of finding the issue. Thank you so much for your patience!

**Comment by dozer8383** [comment]
Thank you so much! I'm willing to wait a long time for this shader, I'm determined to finally get to use a real pathtracing shader. Not sure if it's just because its broken, but as the shader is currently, this is one of the more performant shaders for my computer. I'm getting 30-40 fps most of the time (which is more than can be said for Complementary on High preset).

**Comment by Despatra** [comment]
I changed the format of the textures I listed might be corrupted. let me know if this works
[Phoxel Test.zip](https://github.com/user-attachments/files/16858078/Phoxel.Test.zip)

**Comment by dozer8383** [comment]
I tried the shader you sent, and the results were identical unfortunately. I'm not going to put any more screenshots since it looks the exact same as 0.5.1 (unless you really really want to see it, it's not any different)

**Comment by Despatra** [comment]
Hmm ok, I'm going to send a pack that directly outputs the PT part. I thought that was working, but maybe it's not. I'll get that to you no later than 4 PM today

**Comment by Despatra** [comment]
Ok, TA no longer works but I'm just trying to pin down the issue
[Phoxel Test.zip](https://github.com/user-attachments/files/16900572/Phoxel.Test.zip)

**Comment by dozer8383** [comment]
> Ok, TA no longer works but I'm just trying to pin down the issue [Phoxel Test.zip](https://github.com/user-attachments/files/16900572/Phoxel.Test.zip)

As every good troubleshooter should.

Hopefully these screenshots will help:
![2024-09-05_20 26 55](https://github.com/user-attachments/assets/4afd034b-0694-4b17-afa5-60dc0cf9c24d)
![2024-09-05_20 27 22](https://github.com/user-attachments/assets/b057326f-0c87-43d9-9f0d-da7b08496e53)

**Comment by Despatra** [comment]
Sorry for the wait!, I don't know why but it seems based off of position or view direction? Which is very strange. Just in case I'm going to send you a pack that just doesn't do any PT, make 100% sure it is the PT (which I'm pretty sure it is at this point, but always good to make sure). Unfortunately this is going to be pretty back and forth, wish I knew someone with a similar intel computer so I could do the testing faster. I'll get that to you once again no later than 4 PM MST.

**Comment by dozer8383** [comment]
No problem, man. I've got homework to do anyway.
Do what you need to do to work out the problem. This is slightly fun for me :P

**Comment by Despatra** [comment]
>  This is slightly fun for me :P
Wish I could say the same lol

**Comment by Despatra** [comment]
Here you go:
[Phoxel Test.zip](https://github.com/user-attachments/files/16955279/Phoxel.Test.zip)

**Comment by dozer8383** [comment]
Lol sorry I went silent for a while. I've had a lot of homework and other things to focus on. I'll get back to you on the shader later today.

**Comment by dozer8383** [comment]
How's this looking? I'm noticing the water is a little noisy, probably correct?
![2024-09-12_20 23 51](https://github.com/user-attachments/assets/f9e3cd4a-430f-469d-9475-c239973752f2)

**Comment by Despatra** [comment]
Yes, water is noisy bcus I implement it poorly. Ok so I at least now know what pass to look in.

**Comment by dozer8383** [comment]
Just checking: is there any progress on this particular issue or is it on hold while you continue to develop the shader?

**Comment by Despatra** [comment]
Sorry, I've just had burnout for the past like, 3 weeks, So I haven't gotten much done. I've been feeling better recently so hopefully I can get back on it. I can't really say for certain though.

**Comment by dozer8383** [comment]
Ah. I understand. I know the feeling, I've started many a massively ambitious project and then proceeded to lose interest or get burned out, so it's totally OK. Just come back to it when you feel like it.

**Comment by Despatra** [comment]
Ok, so I think the issue could be how many textures I have bound? To put simply the base limit is 8, but my computer and others can handle 16 maybe?

**Comment by dozer8383** [comment]
Alright then. When things start getting down to the GPU hardware... I don't know what's happening anymore. Also, it should be emphasized that, as opposed to AMD integrated graphics, this is Intel integrated graphics, which is another term for "there is no gpu"
...in all seriousness, I guess I shouldn't expect such a fancy shader to work properly on my laptop. But if you can make it work... I will be pleasantly surprised.

**Comment by Despatra** [comment]
I do understand that integrated graphics are more just using part of the CPU to do graphics and rendering.
And honestly you should be able to, it's just how well is really the question.
But I have to fix this anyway, as many computers, with integrated graphics or not, don't support more than 8, At least that's my understanding. I think you just have a unique case where instead of not running it we get some fun results!
Although this will have to wait until the next update.

**Comment by Despatra** [comment]
Hey. I have a quick question. What Minecraft launcher do you use?

**Comment by dozer8383** [comment]
I use Prism Launcher. That wouldn't make a difference... would it?

**Comment by Despatra** [comment]
Most launchers set RAM limits, which affect your iGPU since it shares ram with the system. I don't know how much RAM my shader uses but it might be worth a try to use the shader while allowing Java to use 8 GB of RAM.

**Comment by dozer8383** [comment]
I'm not totally certain now, but I recall setting the RAM limit to 8 GB before. I can't check it now, since I erased the USB drive with basically all of my stuff on it (intentionally because the USB drive was getting very slow for some reason, probably just being abused day in and day out). But yeah, I'll set it up again and try that, though it might be a while before I can do it. I can probably set it up to 12 GB if necessary.

**Comment by Despatra** [comment]
Most launchers set RAM limits, which affect your iGPU since it shares ram with the system. I don't know how much RAM my shader uses but it might be worth a try to use the shader while allowing Java to use 8 GB of RAM.

**Comment by Despatra** [comment]
Hello again! In case you haven't seen, I will be rewriting the shader. Since you have had issues with it and I really want it to work for you, I will send you parts of it as it is developed so I can catch it before it becomes too hard to track down. Hope you're ok with being a tester until I'm sure we've passed all the issues you currently have.

**Comment by dozer8383** [comment]
Cool! I've been really busy lately, with school and whatnot, but hopefully I can finally find another computer I can test on. I'll probably use my iMac. It's an iMac 14,1, but with 16 GB of RAM installed, and it's been plenty powerful for Minecraft 1.20.1 with all the Create mods installed and other very resource-intensive things.

**Comment by Despatra** [comment]
Hello again! In case you haven't seen, I will be rewriting the shader. Since you have had issues with it and I really want it to work for you, I will send you parts of it as it is developed so I can catch it before it becomes too hard to track down. Hope you're ok with being a tester until I'm sure we've passed all the issues you currently have.

**Comment by dozer8383** [comment]
It's beautiful. /s
Tested on Minecraft 1.21.3 on Ubuntu 24.04 LTS on the same laptop that I had been before, except now I've got about 11 GB of memory allocated because why not. Oh, and I'm using my old 5:4 screen ratio Dell monitor that's kind of 720p but also not. This is with render mode ON, with render mode OFF it is either pure black or pure white, mostly the latter.
![Screenshot from 2024-11-08 16-41-01](https://github.com/user-attachments/assets/30441a87-f9f4-4a6c-96a3-c3cb6e1185b2)

**Comment by Despatra** [comment]
Yes! Beautiful indeed LOL, Well I'm in the process of re-writing it from scratch, quite literally. So like I said, I'll send you tests

**Comment by Despatra** [comment]
Hello! I almost have the first version of the shader for you to test. On my discord I'm gathering a bunch of people with a mixture  of platforms to ensure cross-platform support, it would be easiest for me if you joined there to help me test this. In case you don't know where the link is: https://discord.com/invite/GSF9D3Q2T5

**Comment by dozer8383** [comment]
Sorry, I'm not allowed to have a Discord account. I'm still able to see the channel, I can download files from it, and all that.

**Comment by Despatra** [comment]
Ok, it's fine if discord doesn't work for you, I assume you're Jasper?
Is there some other way you would rather we talk? This isn't part of the issue anymore, and I would prefer that the alpha versions stay private.
I don't know if you can create chats like that on GitHub or not.

**Comment by dozer8383** [comment]
Actually I got around the email verification, so I can view the Discord server again (but I can't have an account). If you put the shader files in the Discord publicly, then that works for me.

**Comment by Despatra** [comment]
So when you say you're not allowed to have a discord account, what do you mean?
I will add you to a group within the server, and there I post the versions.

**Comment by dozer8383** [comment]
That's a temporary "account" that has no email linked to it that doesn't let me post anything.

**Comment by Despatra** [comment]
Oh, ok. I'm fine with you reporting issues about it here. As long as you can download the shader there.
I'll add your temporary account to the shader testers now so you can view that.

**Comment by Despatra** [comment]
Is there a better way to communicate?
You also seem to have to make a new account every so often, and that sounds tedious.
I might be able to send the files here, if I do it through Google drive I think I can unlock it for only your Google account.
Would that be better?

**Comment by dozer8383** [comment]
Google Drive would work great. Yeah it's a little tedious but it works for just being able to observe the Discord.

**Comment by Despatra** [comment]
Alright here's the [drive](https://drive.google.com/file/d/1mNf7FgtwPHEpX7HLLUVGxjNc432uPhcm/view?usp=sharing), hopefully you can request access and I'll open that up available to you.
